### PR TITLE
WIP: Implement systemd-notify protocol

### PR DESCRIPTION
**Pull request by clarfonthey**
Right now, the systemd service will mark gitea as started immediately after the process is started, even though there are multiple time-consuming steps that might be run before the server is actually up. This implements systemd's notify protocol to indicate when the server is ready to receive requests, and when it's shutting down. This is particularly useful for socket activation, since systemd will hold the traffic in a queue until the process is ready, to avoid dropping requests.

Since notifications are done via an environment variable, this has no effect on systems where this environment variable is not present.

Marking as a draft since I'm currently developing from a pinebook and it's a bit too slow for me to test the changes for now, but I wanted to put up the code as a proof of concept and will mark it ready once I test it.

### CODE REVIEW

File: modules/graceful/server.go [hunk 1]
```diff
@@ -152,6 +152,54 @@ func (srv *Server) ListenAndServeTLSConfig(tlsConfig *tls.Config, serve ServeFun
...
+	case StoppingMsg:
+		return []byte("STOPPING=1")
+	}
+}
+
+// systemd notify protocol
+func (srv *Server) Notify(msg NotifyMsg) {
+	notifySocket := Os.Getenv("NOTIFY_SOCKET")
+	if notifySocket == "" {
+		log.Debug("no NOTIFY_SOCKET given")
```
Comment (delvh) [inline modules/graceful/server.go:175]:
```suggestion
		log.Debug("no NOTIFY_SOCKET given, Gitea is probably not deployed using systemd")
```

File: modules/graceful/server.go [hunk 2]
```diff
@@ -152,6 +152,54 @@ func (srv *Server) ListenAndServeTLSConfig(tlsConfig *tls.Config, serve ServeFun
...
+}
+
+func (msg *NotifyMsg) Bytes() []byte {
+	switch msg {
+	case ReadyMsg:
+		return []byte("READY=1")
+	case StoppingMsg:
+		return []byte("STOPPING=1")
+	}
+}
```
Comment (delvh) [inline modules/graceful/server.go:169]: As far as I can see, only the bytes are needed in your implementation.
So, why not skip the ints and make them strings/ byte arrays immediately?

File: modules/graceful/server.go [hunk 3]
```diff
@@ -152,6 +152,54 @@ func (srv *Server) ListenAndServeTLSConfig(tlsConfig *tls.Config, serve ServeFun
...
+
+// systemd notify protocol
```
Comment (delvh) [inline modules/graceful/server.go:171]:
```suggestion
// Notify notifies systemd (if present) about the given startup status of Gitea
```

File: modules/graceful/server.go [hunk 4]
```diff
@@ -152,6 +152,54 @@ func (srv *Server) ListenAndServeTLSConfig(tlsConfig *tls.Config, serve ServeFun
...
+
+	conn, err := net.DialUnix(socketAddr.Net, nil, socketAddr)
+	if err != nil {
+		log.Warn("failed to dial NOTIFY_SOCKET: %v", err)
+		return
+	}
+	defer conn.close()
+
+	if _, err = conn.Write(msg.Bytes()) {
+		log.Warn("failed to notify NOTIFY_SOCKET: %v", err)
```
Comment (delvh) [inline modules/graceful/server.go:198]:
```suggestion
		SystemdNotificationImpossible("unsetenv", err)
		return
	}

	conn, err := net.DialUnix(socketAddr.Net, nil, socketAddr)
	if err != nil {
		SystemdNotificationImpossible("dial", err)
		return
	}
	defer conn.close()

	if _, err = conn.Write(msg.Bytes()) {
		SystemdNotificationImpossible("notify", err)
```

File: modules/graceful/server.go [hunk 5]
```diff
@@ -152,6 +152,54 @@ func (srv *Server) ListenAndServeTLSConfig(tlsConfig *tls.Config, serve ServeFun
...
+
+func (msg *NotifyMsg) Bytes() []byte {
+	switch msg {
+	case ReadyMsg:
+		return []byte("READY=1")
+	case StoppingMsg:
+		return []byte("STOPPING=1")
+	}
+}
+
```
Comment (delvh) [inline modules/graceful/server.go:170]:
```suggestion

func SystemdNotificationImpossible(action string, err error) {
  log.Warn("Failed to %s NOTIFY_SOCKET (%v). Systemd might show some weird states: %v", action, os.Getenv("NOTIFY_SOCKET"), err)
}

```

File: modules/graceful/server.go (hunk 2 above)
Comment (clarfonthey) [reply modules/graceful/server.go:169]: I was trying to pretend that golang has abstract data types but I probably should just give up and use byte strings.

**Comment by zeripath** [comment]
So I think this is in the wrong place. The Notify function will belong in manager_unix.go and manager.go not server.go.

The functions and structs in server.go are called for multiple connections that gitea uses. The manager is the controller and it is that which should be notifying things.

**Comment by clarfonthey** [comment]
Will close this in favour of @zeripath's PR since they're putting the work into this.

**Pull request closed without merging**